### PR TITLE
Lesson 09 - Emmanuel Twum - Mail Room Part 5

### PR DESCRIPTION
**Pull request by etwum**
submitting mail room part 5

### CODE REVIEW

**Comment by natasha-aleksandrova** [comment]
@etwum somehow `.gitignore` file changed in this PR. Please revert that change so that we can merge the PR.

File: students/etwum/lesson09/mailroom_part5.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+        return self.donors
+
+
+class Interaction:
+
+    def __init__(self, donor_list = None):
+        if donor_list is None:
+            donor_list = []
+        else:
+            self.donor_list = donor_list
```
Comment (natasha-aleksandrova) [inline students/etwum/lesson09/mailroom_part5.py]: a bug maybe? if `donor_list` is None then you will never `self.donor_list`

File: students/etwum/lesson09/mailroom_part5.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+    def add_new_donor(self,new_donor):
+        # method to add a new donor
+        self.donor_list.append(new_donor)
+
+    def send_thank_you(self):
```
Comment (natasha-aleksandrova) [inline students/etwum/lesson09/mailroom_part5.py]: this function is a bit too big, would be good to break it up into smaller functions for readability

File: students/etwum/lesson09/mailroom_part5.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            if str_choice in user_selection:
+                # returns a value from user_selection dictionary in the options function
+                # then runs the function from the dictionary
+                options()[str_choice]()
+            elif str_choice == 4:
+                print("\nExiting Program")
+            else:
+                raise Exception
+        except Exception:
+            print("Please input a valid option: 1, 2, 3, or 4")
```
Comment (natasha-aleksandrova) [inline students/etwum/lesson09/mailroom_part5.py]: you could simplify this a bit:
```
try:
    str_choice = input("")  # you can just use strings for `user_selection` -> {"1": run.send_thank_you}
    user_selection[str_choice]()
except KeyError:
    print("Please enter valid options")
```
you could also use `while True` and have '4' be redirected to `sys.exit` function

File: students/etwum/lesson09/mailroom_part5.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+# initial list of donors
+
+
+class DonorData:
+    def __init__(self,donors):
+        self.donors = donors
+
+    def donor_data(self):
+        return self.donors
```
Comment (natasha-aleksandrova) [inline students/etwum/lesson09/mailroom_part5.py]: would be good if your Donor class had `name` attribute as well as total, avg, donations etc basically everything directly related to the Donor object

**Comment by etwum** [comment]
There shouldnt be conflicts in the pull request now

**Comment by natasha-aleksandrova** [comment]
@etwum look at the Files Changed tab :)

**Comment by natasha-aleksandrova** [comment]
@etwum 

what commands did you run?

what does `git remote -v` show?

you may try creating a separate branch from a last good commit:
`git checkout -b master2 <commithash>`

the reason I suggest above is because you have already tried reverting based on your commit history.

if you do create second master branch, then you will need to copy over your new lesson files there.

**Comment by etwum** [comment]
I ran $ git pull upstream master trying to fix the gitignore issue not realizing I could have fixed it an answer way. Would deleting all the other student folders except mine and running the upstream pull again work?

$ git remote -v
origin  https://github.com/etwum/Self_Paced-Online.git (fetch)
origin  https://github.com/etwum/Self_Paced-Online.git (push)
upstream        https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online.git (fetch)
upstream        https://github.com/UWPCE-PythonCert-ClassRepos/Self_Paced-Online.git (push)

**Comment by natasha-aleksandrova** [comment]
I would recommend using commands from that link I sent earlier to keep your fork up to date.

You can try below command as well:
`git rebase upstream/master`

**Comment by natasha-aleksandrova** [comment]
@etwum I could try fixing your repo by cloning your fork?

You would have to add me as collaborator I think so that I can have permissions to push.

**Comment by etwum** [comment]
added

**Comment by natasha-aleksandrova** [comment]
@etwum are you not on slack? it would much easier to talk over chat.

I think I may have a fix for you.

**Comment by etwum** [comment]
I'm not on slack but I can download it. I have another class today but I can talk tomorrow. I'm available anytime after 5:30 pm tomorrow

**Comment by natasha-aleksandrova** [comment]
That would be much easier, I would like to go over the changes I would make before actually making them.

5:30pm should tomorrow should work. Ping me on slack when you're available.

**Comment by etwum** [comment]
what email should i use to invite you to the slack chat? first time i've used slack

**Comment by etwum** [comment]
actually here's the invite link: https://join.slack.com/t/python-helpworkspace/shared_invite/enQtMzQ0ODMzOTcyOTkzLWE1MDkyYTBmZWJhN2I1ZmZiMGI1OWE3YmE4ZTVlZmQ3NDQwYjZjNTliMWE0NDU5NWU3NDQ0NTBkZjg3OWYyZGE

**Comment by natasha-aleksandrova** [comment]
yes message me there. going to remove the link from your comment since this is public repo.

**Comment by etwum** [comment]
how?

**Comment by etwum** [comment]
i see invite people but i need an email

**Comment by natasha-aleksandrova** [comment]
(i'm assuming the link you had was the invite link to our class slack chat)
click the invite link and go through the steps of sign up

**Comment by etwum** [comment]
oh no. didnt know there was a class slack chat

**Comment by natasha-aleksandrova** [comment]
the class slack sign up link was sent via welcome email during sign up and also posted in introductions.

here is the link again, but i will remove it in few mins:
https://join.slack.com/t/na-python-cert/shared_invite/enQtMjkwNzExNzAxNTM5LWMxMWZiODIzYzUzMjI5MzVlOTQ3NTc2YTUzNGFlZDAyN2Y3OWFiYWMxNmE5ZWRkNmZmMzVjNTc1NjgyODcwOTY

**Pull request merged**